Prevent unnecessarily frequent default expunge operations

madmajestro opened the pull request:
A full cache wipe is now performed if too little SHM has been freed by removing expired entries during the default expunge operation. This is intended to prevent unnecessarily frequent calls of the default expunge operation, which should reduce cpu load in some cases. The new configuration setting apc.expunge_threshold defaults to 0.5 (0.5%) and specifies the percentage of SHM that must be available after expired entries have been removed during the default expunge operation. Since apc.expunge_threshold replaces apc.smart, apc.smart has been removed.

nikic commented on php_apc.c:
```diff
@@ -145,7 +145,7 @@ STD_PHP_INI_ENTRY("apc.shm_size",       "32M",  PHP_INI_SYSTEM, OnUpdateShmSize,
 STD_PHP_INI_ENTRY("apc.gc_ttl",         "3600", PHP_INI_SYSTEM, OnUpdateLong,              gc_ttl,           zend_apcu_globals, apcu_globals)
 STD_PHP_INI_ENTRY("apc.ttl",            "0",    PHP_INI_SYSTEM, OnUpdateLong,              ttl,              zend_apcu_globals, apcu_globals)
-STD_PHP_INI_ENTRY("apc.smart",          "0",    PHP_INI_SYSTEM, OnUpdateLong,              smart,            zend_apcu_globals, apcu_globals)
+STD_PHP_INI_ENTRY("apc.smart",          "1",    PHP_INI_SYSTEM, OnUpdateLong,              smart,            zend_apcu_globals, apcu_globals)
```
Probably it's best to not use apc.smart. Nobody really understands what this [undocumented](https://www.php.net/manual/en/apcu.configuration.php) option does...

nikic reviewed (commented):
Having the option specify the value in thousands is a bit weird. Might it make more sense to make this a floating point value?

madmajestro commented:
Fixed. Yes, that's a good idea. It's easier to understand and more flexible.

nikic commented on php_apc.c:
```diff
@@ -89,6 +98,21 @@ static void php_apc_init_globals(zend_apcu_globals* apcu_globals)
+		return FAILURE;
+	}
+
+	double *p = (double *) ZEND_INI_GET_ADDR();
```
Why does this use a different approach (ZEND_INI_GET_ADDR instead of using APCG) than the other ini settings?

nikic reviewed (approved):
LGTM

madmajestro commented:
> I wonder whether it would make sense to create a release from the current master branch before merging this.

I agree. It probably wouldn't hurt, but this is the safer approach.

But please take a look at #606 before creating a release.

madmajestro committed:
Prevent unnecessarily frequent default expunge operations

A full cache wipe is now performed if too little SHM has been freed by
removing expired entries during the default expunge operation. This is
intended to prevent unnecessarily frequent calls of the default expunge
operation, which should reduce cpu load in some cases. The new
configuration setting apc.expunge_threshold defaults to 0.5 (0.5%) and
specifies the percentage of SHM that must be available after expired
entries have been removed during the default expunge operation. Since
apc.expunge_threshold replaces apc.smart, apc.smart has been removed.
